### PR TITLE
 Added directed edges functionality to D3 Visualization

### PR DESCRIPTION
**Pull request by AdamCordner641**
Added DirectedGraph example (VirusOnNetworkDirected).
Appended arrowhead SVG to D3.js module.
Fixes issue https://github.com/projectmesa/mesa/issues/667 (partially).
Based off pull request https://github.com/projectmesa/mesa/pull/846.

### CODE REVIEW

File: mesa/visualization/templates/js/NetworkModule_d3.js [hunk 1]
```diff
@@ -24,6 +24,45 @@ var NetworkModule = function(svg_width, svg_height) {
             g.attr("transform", d3.event.transform);
         }));
 
+    svg.append("svg:defs").selectAll("marker")
```
Comment (rht) [inline mesa/visualization/templates/js/NetworkModule_d3.js:27]: The 2 `svg.append...`'s here are very similar, and so they could be deduplicated to make it clear.

Comment (rht) [reply mesa/visualization/templates/js/NetworkModule_d3.js:27]: There are 3 of them actually.

File: examples/virus_on_network_directed/README.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
+# Virus on a Network
+
+## Summary
+
+This model is based on the NetLogo model "Virus on Network".
```
Comment (rht) [inline examples/virus_on_network_directed/README.md:5]: Needs to say that this model is almost the same as examples/virus_on_network except that the graph is directed. I haven't fully read the code to confirm that this is the only difference.

**Comment by rht** [comment]
In the .git/config of your Mesa git repo, make sure to add
```
[pull]
    rebase = true
```

This will prevent the noisy merge commits when you pull from Mesa main to update your branch.

**Comment by tpike3** [comment]
@AdamCordner641 My impression is you are new to open source development so let us know if you have any questions. Everyday I am learning something new.

File: mesa/visualization/templates/js/NetworkModule_d3.js (hunk 1 above)
Comment (AdamCordner641) [reply mesa/visualization/templates/js/NetworkModule_d3.js]: I've deduplicated the shared code so the changes are as short as possible. Also, the network won't be drawn as directed unless "directed": True in portrayal["edges"]. If "directed": False, or not specified, the network will be drawn as undirected.

File: examples/virus_on_network_directed/README.md (hunk 2 above)
Comment (AdamCordner641) [reply examples/virus_on_network_directed/README.md:5]: I've added some additional text to clarify that the only difference is that the graph is directed.

**Comment by rht** [comment]
I diff-ed the model.py of this PR with the undirected version:
```
18c18
<     return sum(1 for a in model.grid.get_all_cell_contents() if a.state is state)
---
>     return sum([1 for a in model.grid.get_all_cell_contents() if a.state is state])
33c33
< class VirusOnNetwork(Model):
---
> class VirusOnNetworkDirected(Model):
```

1. ` return sum(1 for a in model.grid.get_all_cell_contents() if a.state is state)` is preferred
2. this means that the model graph is actually undirected, because there is no difference. `erdos_renyi_graph` has `directed` as an argument, which is not used

**Comment by rht** [comment]
Here is a diff of server.py:
```
8c8
< from .model import VirusOnNetwork, State, number_infected
---
> from .model import VirusOnNetworkDirected, State, number_infected
37c37,39
<             "tooltip": f"id: {agents[0].unique_id}<br>state: {agents[0].state.name}",
---
>             "tooltip": "id: {}<br>state: {}".format(
>                 agents[0].unique_id, agents[0].state.name
>             ),
47a50,51
>             "directed": True,
>             "marker_size": "large"
68c72
<         ratio_text = "&infin;" if ratio is math.inf else f"{ratio:.2f}"
---
>         ratio_text = "&infin;" if ratio is math.inf else "{0:.2f}".format(ratio)
138c142
<     VirusOnNetwork, [network, MyTextElement(), chart], "Virus Model", model_params
---
>     VirusOnNetworkDirected, [network, MyTextElement(), chart], "Directed Virus Model", model_params
140c144
< server.port = 8521
---
> server.port = 8521
```

The `"...".format(...)` needs to be replaced with the more modern f-strings.

**Comment by rht** [comment]
@AdamCordner641  I'm not sure how it happened, but you have accidentally included a commit in the main branch in this PR. Have you made sure to add https://github.com/projectmesa/mesa/pull/1223#issuecomment-1084010129 ? And then you do `git pull upstream main`, assuming `projectmesa/mesa` remote is called upstream in your .git/config.

File: mesa/visualization/templates/js/NetworkModule_d3.js [hunk 3]
```diff
@@ -24,6 +24,24 @@ var NetworkModule = function(svg_width, svg_height) {
...
+        .attr("viewBox", "0 -5 10 10")
+        .attr("refX", 15)
+        .attr("refY", 0.5)
+        .attr("orient", "auto")
+        .append("svg:path")
+        .attr("d", "M0,-5L10,0L0,5");
+    }
+
+    var marker_sizes = ["end", "endlarge", "endxl"];
+    for (let index = 0; index < marker_sizes.length; index++) {
```
Comment (rht) [inline mesa/visualization/templates/js/NetworkModule_d3.js:40]: Use `for ... of`: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of

File: mesa/visualization/templates/js/NetworkModule_d3.js [hunk 4]
```diff
@@ -24,6 +24,24 @@ var NetworkModule = function(svg_width, svg_height) {
...
+        addMarkers(svg.append("svg:defs").selectAll("marker").data([marker_sizes[index]]));
+        d3.select(`#${marker_sizes[index]}`).attr("markerWidth", 4 + index).attr("markerHeight", 4 + index);
```
Comment (rht) [inline mesa/visualization/templates/js/NetworkModule_d3.js:42]: ? Why is the index necessary here? Also, the 2 lines above seem to be under indented. Use https://github.com/prettier/prettier to auto-format the js file.

Comment (rht) [reply mesa/visualization/templates/js/NetworkModule_d3.js:42]: The whole file seems to not have been formatted properly. Wait until #1225 before you pull and apply prettier to this PR.

Comment (tpike3) [reply mesa/visualization/templates/js/NetworkModule_d3.js:42]: #1225 Merged

Comment (rht) [reply mesa/visualization/templates/js/NetworkModule_d3.js:42]: #1225 has been merged, you need to pull from latest main, and then force push to your branch.

Comment (rht) [reply mesa/visualization/templates/js/NetworkModule_d3.js:42]:
```
# In your fork's main branch
git pull upstream main
git push -f
```
Assuming that this is specified in your .git/config
```
[remote "upstream"]
	url = git@github.com:projectmesa/mesa.git
	fetch = +refs/heads/*:refs/remotes/upstream/*
```

File: mesa/visualization/templates/js/NetworkModule_d3.js (hunk 3 above)
Comment (AdamCordner641) [reply mesa/visualization/templates/js/NetworkModule_d3.js]: Updated to use for ... of

File: mesa/visualization/templates/js/NetworkModule_d3.js (hunk 4 above)
Comment (AdamCordner641) [reply mesa/visualization/templates/js/NetworkModule_d3.js]: Removed use of index with for...of. And applied prettier to PR files.

**Comment by AdamCordner641** [comment]
Pushed and tested fixes which should resolve the issues, but let me know if I missed anything :)

File: examples/virus_on_network_directed/virus_on_network_directed/model.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+def number_susceptible(model):
+    return number_state(model, State.SUSCEPTIBLE)
+
+
+def number_resistant(model):
+    return number_state(model, State.RESISTANT)
+
+
+class VirusOnNetwork(Model):
```
Comment (rht) [inline examples/virus_on_network_directed/virus_on_network_directed/model.py:33]: Needs to be `VirusOnNetworkDirected`.

Comment (rht) [reply examples/virus_on_network_directed/virus_on_network_directed/model.py:33]: Reminder to rename all occurrences of `VirusOnNetwork` inside examples/virus_on_network_directed/.

File: mesa/visualization/templates/js/NetworkModule_d3.js [hunk 6]
```diff
@@ -80,6 +109,13 @@ var NetworkModule = function (svg_width, svg_height) {
       })
       .attr("stroke", function (d) {
         return d.color;
+      })
+      .attr("marker-end", function (d) {
+        return d.directed
+          ? "url(#end" +
+              (d.marker_size !== undefined ? d.marker_size : "") +
```
Comment (rht) [inline mesa/visualization/templates/js/NetworkModule_d3.js:116]: You can do this instead
`(d.marker_size ?? "")`

See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator

**Comment by rht** [comment]
I still think that the example in this PR doesn't showcase directed networks. It's behaviorally identical to the undirected version.
To make it different, I think we need to implement `get_neighbors` in a new class called `DirectedNetworkSpace` [1] in such a way that it returns either of:
- all in and out neighbors
- all in neighbors only
- all out neighbors only

[1] let's call it "space" instead of "grid"

**Comment by rht** [comment]
I think we should defer this PR, and do a separate PR that implements `DirectedNetworkSpace` with tests. Once this is done, we can revisit this PR.

**Comment by tpike3** [comment]
> I think we should defer this PR, and do a separate PR that implements `DirectedNetworkSpace` with tests. Once this is done, we can revisit this PR.

I concur. @AdamCordner641 Thank you very much for all the hard work you have put into this. Your work definitely revealed some work we need to do. I hope you can use the code you developed in your project and would be happy for any assistance you could provide on building out a `DirectedNetworkSpace.`

File: examples/virus_on_network_directed/virus_on_network_directed/model.py (hunk 5 above)
Comment (AdamCordner641) [reply examples/virus_on_network_directed/virus_on_network_directed/model.py]: All occurrences should now be renamed

**Comment by AdamCordner641** [comment]
> > I think we should defer this PR, and do a separate PR that implements `DirectedNetworkSpace` with tests. Once this is done, we can revisit this PR.
> 
> I concur. @AdamCordner641 Thank you very much for all the hard work you have put into this. Your work definitely revealed some work we need to do. I hope you can use the code you developed in your project and would be happy for any assistance you could provide on building out a `DirectedNetworkSpace.`

No problem! Thank you both for helping with PR feedback. I can definitely use these changes I have made in my project. I would also like to be able to use the `get_neighbors` functionality rht described in my project, so I'll try to make time to have a go at implementing a `DirectedNetworkSpace` class.

**Comment by Corvince** [comment]
Thank you for the PR and sorry for never putting it forward

If this is still actively followed please re-open the PR at https://github.com/projectmesa/mesa-viz-tornado
But note that we are moving towards a new pure python frontend

**Pull request closed without merging**